source/h1_connection.c

```diff
@@ -518,6 +542,165 @@ struct aws_http_stream *s_new_client_request_stream(const struct aws_http_reques
+    }
+
+    /* reponse-line: "{version} {status} {status_text}\r\n" */
+    size_t reponse_line_len = 4; /* 2 spaces + "\r\n" */
+    err |= aws_add_size_checked(version.len, reponse_line_len, &reponse_line_len);
+    err |= aws_add_size_checked(status_code.len, reponse_line_len, &reponse_line_len);
+    err |= aws_add_size_checked(status_text.len, reponse_line_len, &reponse_line_len);
+
+    /* head-end: "\r\n" */
+    size_t head_end_len = 2;
+    size_t head_total_len = reponse_line_len;
+    err |= aws_add_size_checked(header_lines_len, head_total_len, &head_total_len);
+    err |= aws_add_size_checked(head_end_len, head_total_len, &head_total_len);
+
+    if (err) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM,
+            "id=%p: Failed to create response, size calculation had error %d (%s).",
+            (void *)stream,
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        return AWS_OP_ERR;
+    }
+
+    err = aws_byte_buf_init(&stream->outgoing_head_buf, stream->base.alloc, head_total_len);
+    if (err) {
+        AWS_LOGF_ERROR(
+            AWS_LS_HTTP_STREAM,
+            "id=%p: Failed to create request, buffer initialization had error %d (%s).",
+            (void *)stream,
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+
+        return AWS_OP_ERR;
+    }
+
+    bool wrote_all = true;
+
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(&stream->outgoing_head_buf, version);
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, ' ');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(&stream->outgoing_head_buf, status_code);
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, ' ');
+    wrote_all &= aws_byte_buf_write_from_whole_cursor(&stream->outgoing_head_buf, status_text);
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, '\r');
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, '\n');
+
+    s_write_headers(&stream->outgoing_head_buf, options->header_array, options->num_headers);
+
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, '\r');
+    wrote_all &= aws_byte_buf_write_u8(&stream->outgoing_head_buf, '\n');
+    (void)wrote_all;
+    AWS_ASSERT(wrote_all);
+
+    /* Success! */
+    return AWS_OP_SUCCESS;
+}
+
+static int s_stream_send_response(struct aws_http_stream *stream, const struct aws_http_response_options *options) {
+    struct h1_stream *h1_stream = AWS_CONTAINER_OF(stream, struct h1_stream, base);
+    struct h1_connection *connection = AWS_CONTAINER_OF(stream->owning_connection, struct h1_connection, base);
+    int err;
+    bool should_schedule_task = false;
+    { /* BEGIN CRITICAL SECTION */
+        err = aws_mutex_lock(&connection->synced_data.lock);
+        if (h1_stream->synced_data.has_response) {
+            AWS_LOGF_ERROR(AWS_LS_HTTP_CONNECTION, "id=%p: Response already created on the stream", (void *)h1_stream);
```

Comment:
Other places we're logging `id` as the `aws_http_stream *`, which in this case is `stream`, not `h1_stream`.